README.md
![license](https://img.shields.io/github/license/loliot/lot-odroid-n2)
![version](https://img.shields.io/github/v/tag/loliot/lot-odroid-n2?sort=semver)
![language](https://img.shields.io/github/languages/top/loliot/lot-odroid-n2)

# lot-odroid-n2

## Installation

<strong><a href="https://github.com/loliot/lot#installation" target=_blank>lot installation guide</a></strong>

```bash
sudo lot install cpp
```

## Blink example

```cpp
#include <lot/lot.h>
#include <lot/Gpio.h>

const int LED_PIN = 13;

int main( void )
{
    lot::Gpio led( LED_PIN );
    led.mode( lot::DOUT );

    for( ;; )
    {
        led.toggle();
        lot::delay_ms( 200 );
    }
}
```

```bash
g++ blink.cpp -o blink -llot &&\
./blink
```


debian/changelog
lot-odroid-n2 (0.14.3) unstable; urgency=medium

  * Rename enum identifier to mixed case without '_'

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 30 Jan 2020 20:09:42 +0900

lot-odroid-n2 (0.14.2) unstable; urgency=medium

  * Add virtual destructor
  * Add override keyword
  * Move functions using cs_pin into ISpi.hpp
  * Add using-declaration to introduce members of IGpio

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 29 Jan 2020 14:24:21 +0900

lot-odroid-n2 (0.14.1) unstable; urgency=medium

  * Use -O2 instead of -O1 and remove -g

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 23 Jan 2020 12:10:25 +0900

lot-odroid-n2 (0.14.0) unstable; urgency=medium

  * Correct parameter order
  * Move init() to constructor

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 18 Jan 2020 01:19:52 +0900

lot-odroid-n2 (0.13.5) unstable; urgency=medium

  * Change type of 'start_millis/micros' to 'std::atomic'

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 17 Jan 2020 16:53:21 +0900

lot-odroid-n2 (0.13.4) unstable; urgency=medium

  * Refactor Gpio class
  * Update docs

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 16 Jan 2020 12:28:54 +0900

lot-odroid-n2 (0.13.3) unstable; urgency=medium

  * Fix incorrect pull_up_down register access

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 10 Jan 2020 14:48:41 +0900

lot-odroid-n2 (0.13.2) unstable; urgency=medium

  * API: Move typedef into loddef.h

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 08 Jan 2020 10:38:21 +0900

lot-odroid-n2 (0.13.1) unstable; urgency=medium

  * API: Fix bug where mode() does not set m_mode

 -- Hyeonki Hong <hhk7734@gmail.com>  Mon, 06 Jan 2020 17:16:47 +0900

lot-odroid-n2 (0.13.0) unstable; urgency=medium

  * API: Refactor type definition
  * API: Update license
  * API: Remove LOT mode variables
  * API: Remove set/get_pin_speed
  * API: Remove set_, get_, _write, _read
  * API: Move GPIO functions into gpio namespace
  * API: Implement Gpio class
  * API: Add xx_stoi, xx_itos in lotdef.h
  * Implement analog()
  * Remove gpio()

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 05 Jan 2020 22:49:54 +0900

lot-odroid-n2 (0.12.0) unstable; urgency=medium

  * Implement unsupported_error class
  * Implement get/set_pin_drive()

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 02 Jan 2020 12:04:34 +0900

lot-odroid-n2 (0.11.0) unstable; urgency=medium

  * Implement get_lot_pin_available(pin)
  * Use throw to handle errors instead of exit()

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 01 Jan 2020 23:32:47 +0900

lot-odroid-n2 (0.10.2) unstable; urgency=medium

  * Fix debian install error

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 01 Jan 2020 00:00:02 +0900

lot-odroid-n2 (0.10.1) unstable; urgency=medium

  * API: Add pkg-config file
  * API: Refactor Log

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 31 Dec 2019 17:21:34 +0900

lot-odroid-n2 (0.10.0) unstable; urgency=medium

  * Implement get_pin_mode()
  * Implement get/set_pull_up_down()

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 29 Dec 2019 18:01:59 +0900

lot-odroid-n2 (0.9.0) unstable; urgency=medium

  * API: Add INFO log level
  * API: Move *_linux.cpp into linux directory
  * API: Prevent reopen when already open
  * Implement test code

 -- Hyeonki Hong <hhk7734@gmail.com>  Sun, 29 Dec 2019 15:16:03 +0900

lot-odroid-n2 (0.8.3) unstable; urgency=medium

  * API: Set default argument
  * API: Add checking if pin is available
    Pin mux was not considered completely.
  * Correct variable name

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 24 Dec 2019 02:33:03 +0900

lot-odroid-n2 (0.8.2) unstable; urgency=medium

  * API: Remove infinite loop
  * Add log message in lot_gpio.cpp

 -- Hyeonki Hong <hhk7734@gmail.com>  Mon, 18 Nov 2019 22:30:41 +0900

lot-odroid-n2 (0.8.1) unstable; urgency=medium

  * Add delay_us_2() (delay using hard loop)

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 09 Nov 2019 14:10:40 +0000

lot-odroid-n2 (0.8.0) unstable; urgency=medium

  * API: Add lot_time_init()
  * API: Implement lot_time for Linux

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 09 Nov 2019 12:58:13 +0000

lot-odroid-n2 (0.7.0) unstable; urgency=medium

  * API: Implement Log

 -- Hyeonki Hong <hhk7734@gmail.com>  Fri, 08 Nov 2019 22:25:23 +0900

lot-odroid-n2 (0.6.0) unstable; urgency=medium

  * API: Implement SPI source for Linux
  * API: Add source template
  * Use Spi_linux.cpp for odroid-n2

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 31 Oct 2019 14:35:36 +0000

lot-odroid-n2 (0.5.0) unstable; urgency=medium

  * API: Implement I2C source for Linux
  * Use I2c_linux.cpp for odroid-n2

 -- Hyeonki Hong <hhk7734@gmail.com>  Tue, 22 Oct 2019 23:49:38 +0900

lot-odroid-n2 (0.4.0) unstable; urgency=medium

  * API: Implement ios, ostream class
  * API: Implement UART source for Linux
  * Use Uart_linux.cpp for odroid-n2

 -- Hyeonki Hong <hhk7734@gmail.com>  Sat, 12 Oct 2019 15:28:54 +0900

lot-odroid-n2 (0.3.1) unstable; urgency=medium

  * API: Test debian package
  * Add basic reg-map and pin-map
  * Add basic digital read/write function

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 06 Oct 2019 09:03:00 +0900

lot-odroid-n2 (0.3.0) unstable; urgency=medium

  * API: Test Makefile
  * API: Change source tree
  * API: Add I2C, UART, SPI header

 -- Hyeonki Hong <hhk7734@gmail.com>  Wed, 03 Oct 2019 13:50:00 +0900

lot-odroid-n2 (0.2.0) unstable; urgency=medium

  * API: Add GPIO header, UART interface, Doxyfile

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 26 Sep 2019 14:48:00 +0900

lot-odroid-n2 (0.1.0) unstable; urgency=medium

  * API: Initial release.
  * API: Add I2C, SPI, time interface

 -- Hyeonki Hong <hhk7734@gmail.com>  Thu, 22 Sep 2019 22:53:00 +0900


lot_SBC.h
/*
 * MIT License
 * Copyright (c) 2019-2020 Hyeonki Hong <hhk7734@gmail.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "s922x.h"
#include "lot-API/lotdef.h"

namespace lot
{
constexpr int LAST_PHY_PIN = 40;

const int phy_to_gpio[LAST_PHY_PIN + 1] = {
    // clang-format off
     UNUSED ,
     UNUSED , UNUSED ,   // 3.3V                 1   | 2     5.0V
     GPIOX17, UNUSED ,   // GPIOX.17(I2C-2_SDA)  3   | 4     5.0V
     GPIOX18, UNUSED ,   // GPIOX.18(I2C-2_SCL)  5   | 6     GND
     GPIOA13, GPIOX12,   // GPIOA.13             7   | 8     GPIOX.12(UART_TX_B)
     UNUSED , GPIOX13,   // GND                  9   | 10    GPIOX.13(UART_RX_B)
     GPIOX3 , GPIOX16,   // GPIOX.3              11  | 12    GPIOX.16(PWM_E)
     GPIOX4 , UNUSED ,   // GPIOX.4              13  | 14    GND
     GPIOX7 , GPIOX0 ,   // GPIOX.7(PWM_F)       15  | 16    GPIOX.0
     UNUSED , GPIOX1 ,   // 3.3V                 17  | 18    GPIOX.1
     GPIOX8 , UNUSED ,   // GPIOX.8(SPI_MOSI)    19  | 20    GND
     GPIOX9 , GPIOX2 ,   // GPIOX.9(SPI_MISO)    21  | 22    GPIOX.2
     GPIOX11, GPIOX10,   // GPIOX.11(SPI_SCLK)   23  | 24    GPIOX.10(SPI_CE0)
     UNUSED , GPIOA4 ,   // GND                  25  | 26    GPIOA.4(SPI_CE1)
     GPIOA14, GPIOA15,   // GPIOA.14(I2C-3_SDA)  27  | 28    GPIOA.15(I2C-3_SCL)
     GPIOX14, UNUSED ,   // GPIOX.14             29  | 30    GND
     GPIOX15, GPIOA12,   // GPIOX.15             31  | 32    GPIOA.12
     GPIOX5 , UNUSED ,   // GPIOX.5(PWM_C)       33  | 34    GND
     GPIOX6 , GPIOX19,   // GPIOX.6(PWM_D)       35  | 36    GPIOX.19
     UNUSED , UNUSED ,   // ADC.AIN3             37  | 38    1.8V REF OUT
     UNUSED , UNUSED     // GND                  39  | 40    ADC.AIN2
    // clang-format on
};

const bool is_available_phy[LAST_PHY_PIN + 1] = {
    // clang-format off
     false,
     false, false,   // 3.3V                 1   | 2     5.0V
     true , false,   // GPIOX.17(I2C-2_SDA)  3   | 4     5.0V
     true , false,   // GPIOX.18(I2C-2_SCL)  5   | 6     GND
     true , true ,   // GPIOA.13             7   | 8     GPIOX.12(UART_TX_B)
     false, true ,   // GND                  9   | 10    GPIOX.13(UART_RX_B)
     true , true ,   // GPIOX.3              11  | 12    GPIOX.16(PWM_E)
     true , false,   // GPIOX.4              13  | 14    GND
     true , true ,   // GPIOX.7(PWM_F)       15  | 16    GPIOX.0
     false, true ,   // 3.3V                 17  | 18    GPIOX.1
     true , false,   // GPIOX.8(SPI_MOSI)    19  | 20    GND
     true , true ,   // GPIOX.9(SPI_MISO)    21  | 22    GPIOX.2
     true , true ,   // GPIOX.11(SPI_SCLK)   23  | 24    GPIOX.10(SPI_CE0)
     false, true ,   // GND                  25  | 26    GPIOA.4(SPI_CE1)
     true , true ,   // GPIOA.14(I2C-3_SDA)  27  | 28    GPIOA.15(I2C-3_SCL)
     true , false,   // GPIOX.14             29  | 30    GND
     true , true ,   // GPIOX.15             31  | 32    GPIOA.12
     true , false,   // GPIOX.5(PWM_C)       33  | 34    GND
     true , true ,   // GPIOX.6(PWM_D)       35  | 36    GPIOX.19
     false, false,   // ADC.AIN3             37  | 38    1.8V REF OUT
     false, false    // GND                  39  | 40    ADC.AIN2
    // clang-format on
};
}    // namespace lot